### ISSUE TITLE
Rework exceptions

### ISSUE THREAD

**Issue by jage** [twingly/twingly-url#31]
First of, we shouldn't raise that many exceptions if we can avoid it. I've made some comments on what we should return on exception here below:
- `Twingly::URL.parse` - Returns a Struct with `#url` and `#domain` accessors: _Return "empty" struct on exception_
- `Twingly::URL.validate` - Validates a URL: _Return false on exception_
- `Twingly::URL::Normalizer.normalize(string)` - Extracts URLs from string (Array): _Return empty array on exception_
- `Twingly::URL::Utilities.remove_scheme(url)` - Removes scheme from HTTP/HTTPS URLs (`http://twingly.com` -> `//twingly.com`): _Always run to_str on input, always remove scheme (if there's nothing to remove, return same input as string)_

Not sure about the hashers: _Run to_str on input (won't accept nil etc), always return hash?_
- `Twingly::URL::Hasher.taskdb_hash(url)` - MD5 hexdigest:
- `Twingly::URL::Hasher.blogstream_hash(url)` - MD5 hexdigest
- `Twingly::URL::Hasher.documentdb_hash(url)` - SHA256 unsigned long, native endian digest
- `Twingly::URL::Hasher.autopingdb_hash(url)` - SHA256 64-bit signed, native endian digest
- `Twingly::URL::Hasher.pingloggerdb_hash(url)` - SHA256 64-bit unsigned, native endian digest

But if we happen to leak an exception from the libraries used in this gem, they should be tagged by a `Twingly::URL` error module to make it easier on the user to handle.

``` Ruby
# Pseudo code
module Twingly::URL
  module Error
  end
end

# ...

def self.normalize_url
  # Do some fancy stuff
rescue Exception => error
  error.extend(Twingly::URL::Error)
  raise
end
```

**Comment by jage** [comment]
As an alternative to module tagging we could wrap the exceptions, I like how Sequel does it:

* Rescue errors, pass to `raise_error` https://github.com/jeremyevans/sequel/blob/4e222141649eb610bf5bdfd59f117a98e364e8c5/lib/sequel/adapters/mysql2.rb#L72-L100
* `raise_error` converts exception class: https://github.com/jeremyevans/sequel/blob/9cc1ba7ecc682fe2c5af4c876356806f83ce0b02/lib/sequel/database/misc.rb#L419-L427
* `convert_exception_class` stores the wrapped exception: https://github.com/jeremyevans/sequel/blob/4e222141649eb610bf5bdfd59f117a98e364e8c5/lib/sequel/core.rb#L104-L113

There are some logic for keeping it database independant that we might not need.

**Comment by dentarg** [comment]
I looked at the links about what Sequel does, but it was quite hard to follow everything (mostly the use of `opts[:classes]` and to understand when Sequel think it's appropriate to convert).

* specs that gave some insight: [spec/core/database_spec.rb](https://github.com/jeremyevans/sequel/blob/4e222141649eb610bf5bdfd59f117a98e364e8c5/spec/core/database_spec.rb#L1737-L1760)
* `database_error_regexps` sample: [lib/sequel/adapters/shared/mysql.rb](https://github.com/jeremyevans/sequel/blob/4e222141649eb610bf5bdfd59f117a98e364e8c5/lib/sequel/adapters/shared/mysql.rb#L388-L396)
* example of calling `raise_error` with `:classes` set: [lib/sequel/database/transactions.rb](https://github.com/jeremyevans/sequel/blob/4e222141649eb610bf5bdfd59f117a98e364e8c5/lib/sequel/database/transactions.rb#L147)

**Comment by dentarg** [comment]
I think I'm happy with just extending the error (#106).

**Comment by dentarg** [comment]
As `Hasher` might be removed, https://github.com/twingly/twingly-url/issues/77, let's keep `Hasher` out of this. Then I think everything here has been addressed (or removed from the gem).
